epix/simulator/helpers.py

```diff
@@ -0,0 +1,116 @@
+import wfsim
+from scipy.interpolate import interp1d
+import numpy as np
+import numba
+
+#Numba and classes still are not a match made in heaven
+@numba.njit
+def _merge_these_clusters(s2_area1, z1, s2_area2, z2):
+    sensitive_volume_ztop=0 #it's the ground mesh, the top liquid level is at 2.7; // mm
+    max_s2_area = max(s2_area1,s2_area2)
+    if max_s2_area>5000:
+        SeparationDistanceIntercept = 0.00024787 * 5000. + 3.4056346550312973
+        SeparationDistanceSlope = 5.5869678412887262e-07 * 5000. + 0.0044792968
+    else:
+        SeparationDistanceIntercept = \
+          0.00024787 * max_s2_area + 3.4056346550312973
+        SeparationDistanceSlope = \
+          5.5869678412887262e-07 * max_s2_area + 0.0044792968
+    SeparationDistance = \
+        SeparationDistanceIntercept - \
+        SeparationDistanceSlope * (-sensitive_volume_ztop +\
+            (z1 + z2) * 0.5)
+    return z1-z2<SeparationDistance
+
+class Helpers():
+    @staticmethod
+    def assignOrder(order):
+        """This is a trick to have the calculation function be executed in a particular order"""
+        def do_assignment(to_func):
+            to_func.order = order
+            return to_func
+
+        return do_assignment
+
+    @staticmethod
+    def average_spe_distribution(spe_shapes):
+        """We take the spe distribution from all channels and take the average to be our spe distribution to draw photon areas from"""
+        uniform_to_pe_arr = []
+        for ch in spe_shapes.columns[1:]:  # skip the first element which is the 'charge' header
+            if spe_shapes[ch].sum() > 0:
+                # mean_spe = (spe_shapes['charge'].values * spe_shapes[ch]).sum() / spe_shapes[ch].sum()
+                scaled_bins = spe_shapes['charge'].values  # / mean_spe
+                cdf = np.cumsum(spe_shapes[ch]) / np.sum(spe_shapes[ch])
+            else:
+                # if sum is 0, just make some dummy axes to pass to interpolator
+                cdf = np.linspace(0, 1, 10)
+                scaled_bins = np.zeros_like(cdf)
+
+            grid_cdf = np.linspace(0, 1, 2001)
+            grid_scale = interp1d(cdf, scaled_bins,
+                                  bounds_error=False,
+                                  fill_value=(scaled_bins[0], scaled_bins[-1]))(grid_cdf)
+
+            uniform_to_pe_arr.append(grid_scale)
+        spe_distribution = np.mean(uniform_to_pe_arr, axis=0)
+        return spe_distribution
+    
+    @staticmethod
+    @numba.njit
+    def macro_cluster_events(instructions):
+        """Loops over all instructions, checks if it's an s2 and if there is another s2 within the same event
+            within the macro cluster distance, if it is they are merged."""
+        for ix1,_ in enumerate(instructions):
+            if instructions[ix1]['type']!=2:
+                continue
+            for ix2 in range(1,len(instructions[ix1:])+1):
```

Comment:
I am not totally sure if this iterative 1:1 clustering is correct. We could get different results depending on the way we iterate over the instructions. I tried to illustrate this with the sketch below. ![image](https://user-images.githubusercontent.com/27280678/124576227-64812c80-de4c-11eb-94b2-666d275cf2e9.jpeg)